doc/gold.rst

```diff
@@ -0,0 +1,412 @@
+  ASCII characters).
+
+Gold-syntax tokens are delimited by whitespace, which is discarded, and the
+special delimiters, which are considered single tokens. Note that new line
+characters are not treated specially. Also note that comments are treated
+specially by the parser; see below.
+
+A token beginning with a digit in ``0-9`` is assumed to be a numeric value. A
+valid numeric value must be one of the following:
+
+1. A sequence of characters in ``0-9``, such as ``1``, ``42``, or ``094``,
+   denotes a decimal (base-10) integer. Note that numbers beginning with a ``0``
+   are considered to be decimal, not octal.
+
+2. The characters ``0x`` or ``0X`` (a digit zero followed by an uppercase or
+   lowercase letter 'X') followed by a sequence of characters in ``0-9A-Fa-f``,
+   such as ``0x1``, ``0xa9``, or ``0xbeef``, denotes a hexadecimal (base-16)
+   integer.
+
+3. The characters ``0o`` or ``0O`` (a digit zero followed by an uppercase or
+   lowercase letter 'O') followed by a sequence of characters in ``0-7``, such
+   as ``0o644``, denotes an octal (base-8) integer. Note that numbers beginning
+   with a ``0`` (zero) followed by digits are considered to be decimal, not
+   octal.
+
+4. The characters ``0b`` or ``0B`` (a digit zero followed by an uppercase or
+   lowercase letter 'B') followed by a sequence of characters in ``0 1``, such
+   as ``0b11010001``, denotes a binary (base-2) integer.
+
+A token beginning with a digit in ``0-9`` and not conforming to one of the above
+rules is an error.
+
+A token beginning with a letter in ``A-Za-z`` is assumed to be an identifier.
+The restriction to a letter applies only to the first character; identifiers may
+contain any printable ASCII character which does not delimit a token. For
+example, ``foo-bar?`` is considered a valid identifier.
+
+A token beginning with a punctuation character continues until the first
+non-punctuation character, at which point it will end, regardless of whether a
+whitespace character, comment, or special delimiter has been encountered.
+Therefore ``-1`` becomes two tokens.
+
+.. attention:: The above rules imply that the string ``1+2`` is considered a
+               single, invalid, numeric token, and the string ``x-2`` is
+               considered an identifier. Make sure to put spaces around your
+               operators!
+
+
+Language Elements
+=================
+
+A Gold-syntax unit consists of a series of top-level statements.
+
+Binding Statements
+------------------
+
+The following statements introduce bindings of names. They are all allowed at
+the top level of the file, and a subset are allowed inside executable code
+blocks.
+
+Top-level bindings are visible throughout the program. Block-level bindings are
+visible until the end of the scope in which they are introduced. Introducing a
+block-level binding with the same name as another binding, even in the same
+scope, will result in the previous binding being shadowed until the new binding
```

Comment:
"even in the same scope"

why do we want to allow this?

Reply:
That *is* a question. Probably one that I should have actually asked you before just typing it into the document.

Rust allows stuff like this:

```rust
let foo = 7;
let bar = &foo;
let foo = 8;

// foo == 8
// *foo == 7
```

The important part is that a `let` without a `mut` doesn't mean that the *identifier* can't be rebound; just that the section of memory it allocates has the same value for the lifetime of the scope, i.e. if you take a pointer to a non-`mut` binding, the value pointed to by that pointer will not change as long as the pointer is valid. This allows you to re-use a name (like you would in C) without clobbering references to the value.

On the other hand, C# disallows variable shadowing entirely, which is another valid option, but occasionally results in having to come up with a series of creative names for things that you'd rather just use the same name for.

I'm willing to go either way; or alternatively, go the (more conservative) C# route, and simply remove the additional check and switch to the way Rust does it if the C# way turns out to be a pain.

Reply:
oh, okay. i see the value and i like the shadowing option.